File: irtoy/irtoy.py
#!/usr/bin/env python
#
# Class for simplifying the reading and transmitting of IR signals from the IR Toy.
# This only works for firmware revision 22 or greater.
# see https://github.com/crleblanc/PyIrToy and
# http://dangerousprototypes.com/docs/USB_Infrared_Toy for more info.
#
# Chris LeBlanc, 2012
#
#--
#
# This work is free: you can redistribute it and/or modify it under the terms
# of Creative Commons Attribution ShareAlike license v3.0
#
# This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
# without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the License for more details. You should have received a copy of the License along
# with this program. If not, see <http://creativecommons.org/licenses/by-sa/3.0/>.

import sys
import time
import binascii
import logging

__author__ = 'Chris LeBlanc'
__version__ = '0.2.1'
__email__ = 'crleblanc@gmail.com'

class IrToy(object):

    def __init__(self, serialDevice):
        self.toy = serialDevice
        logging.basicConfig(stream=sys.stderr, level=logging.DEBUG)
        # ir toy must be in sampling mode, even for transmissions to work
        self._setSamplingMode()

    def _setSamplingMode(self):
        '''set the IR Toy to use sampling mode, which we use exclusively'''
        self.reset()
        logging.debug('reset')
        self.write(b'S')
        logging.debug('wrote sampling mode')
        protocolVersion = self.toy.read(3)
        if protocolVersion != 'S01':
          raise IOError('Expected protocol "S01", got %s' % protocolVersion)
        logging.debug('got protocol %s' % protocolVersion)

    def _acknowledge(self, code):
        byteCode = bytearray(code)
        for idx in range(0, len(code), 62):
          logging.debug('sending %r', code[idx:idx+62])
          self.write(byteCode[idx:idx+62])
          ack = ord(self.toy.read(1))
          if ack != 0x3e:
            raise IOError('invalid acknowledgement received %d' % ack)


    def reset(self):
      logging.debug('resetting')
      self.write(bytearray([0x00] * 5))

    def write(self, arr):
      # time.sleep(0.05)
      self.toy.write(arr)

    def transmit(self, code):
        if len(code) % 2 != 0:
            raise ValueError("Length of code argument must be an even number")

        # ensure the last two codes are always 0xff (255) to tell the IR Toy
        # it's the end of the signal
        if code[-2:] != [0xff, 0xff]:
            code.extend([0xff, 0xff])

        # Enable transmit handshake, notify on complete, byte count report, and
        # then sent the actual transmit command
        self._acknowledge([0x26, 0x25, 0x24, 0x03])
        logging.debug('sent transmit')

        # Send the actual data (chunked inside)
        self._acknowledge(code)
        logging.debug('sent code')

        # Acknowledge the data was actually sent, and all of it was sent.
        if self.toy.read(1) != 't':
          raise IOError('expected a "t" acknowledgement')
        hexBytes = binascii.b2a_hex(self.toy.read(2))
        if self.toy.read(1) != 'C':
          raise IOError('expected a completion "C"')
        if int(hexBytes, 16) != len(code):
          raise IOError('short write')

        # self._setSamplingMode()

File: server/irtoy.go
package server

import "io"
// import "log"

type IrToy struct {
  serial io.ReadWriteCloser
}

func NewToy(s io.ReadWriteCloser) *IrToy {
  t := &IrToy { serial: s }
  t.setSamplingMode()
  return t
}

func (t *IrToy) setSamplingMode() {
  t.reset()
  t.write([]byte{'S'})
  var version [3]byte
  t.read(version[0:3])
  if version[0] != 'S' || version[1] != '0' || version[2] != '1' {
    panic("invalid protocol version")
  }
}

func (t *IrToy) write(b []byte) {
  // log.Printf("writing %v", b)
  n, err := t.serial.Write(b)
  if err != nil { panic(err) }
  if n != len(b) { panic("short write") }
}

func (t *IrToy) read(b []byte) {
  // log.Printf("waiting for %d bytes", len(b))
  n, err := t.serial.Read(b)
  // log.Printf("got %d bytes: %v", n, b)
  if err != nil { panic(err) }
  if n != len(b) { panic("short read") }
}

func (t *IrToy) reset() {
  t.write([]byte{0, 0, 0, 0, 0})
}

func (t *IrToy) transmit(b []byte) {
  var ack [3]byte
  // Enable handshakes, transmit, and whatnot. The acknowledgement is the size
  // of the internal buffer (amount of bytes we can send)
  t.write([]byte{0x26, 0x25, 0x24, 0x03})
  t.read(ack[0:1])
  size := int(ack[0])
  if size < 0 { panic("bad buffer size") }

  // Send all the data in the buffer-size chunks, reading off acknowledgements.
  for i := 0; i < len(b); i += size {
    end := i + size
    if end > len(b) { end = len(b) }
    t.write(b[i:end])
    if end == len(b) {
      t.write([]byte{0xff, 0xff})
    }

    t.read(ack[0:1])
    if int(ack[0]) != size {
      panic("bad acknowledgement")
    }
  }

  // Finally read the transmit count and completion flags
  t.read(ack[0:3])
  if ack[0] != 't' { panic("didn't receive a 't'") }
  if (int(ack[1]) << 8) | int(ack[2]) != len(b) + 2 {
    panic("didn't send all bytes?")
  }
  t.read(ack[0:1])
  if ack[0] != 'C' { panic("didn't actually complete") }
}
